torchani/data/new.py

```diff
@@ -0,0 +1,458 @@
+            for example: 'HCNO' means {'H': 0, 'C': 1, 'N': 2, 'O': 3}
+        subtract_self_energies (bool): whether subtract self energies from energies
+        self_energies (list): if `subtract_self_energies` is True, the order should keep
+            the same as `species_order`.
+            for example :[-0.600953, -38.08316, -54.707756, -75.194466] will be converted
+            to {'H': -0.600953, 'C': -38.08316, 'N': -54.707756, 'O': -75.194466}
+
+    Returns:
+        ([chunk1, chunk2, ...], {"energies", "force", ...}) in which chunk1=(species, coordinates)
+        e.g. chunk1 = [[1807, 21], [1807, 21, 3]], chunk2 = [[193, 50], [193, 50, 3]]
+        'energies' = [2000, 1]
+    """
+    def __init__(self, file_path,
+                 batch_size=1000,
+                 device='cpu',
+                 chunk_threshold=20,
+                 species_order='HCNO',
+                 subtract_self_energies=False,
+                 self_energies=[-0.600953, -38.08316, -54.707756, -75.194466]):
+
+        super(CachedDataset, self).__init__()
+
+        # example of species_dict will looks like
+        # species_dict: {'H': 0, 'C': 1, 'N': 2, 'O': 3}
+        # self_energies_dict: {'H': -0.600953, 'C': -38.08316, 'N': -54.707756, 'O': -75.194466}
+        species_dict = {}
+        self_energies_dict = {}
+        for i, s in enumerate(species_order):
+            species_dict[s] = i
+            self_energies_dict[s] = self_energies[i]
+
+        self.data_species = []
+        self.data_coordinates = []
+        self.data_energies = []
+        self.data_self_energies = []
+
+        anidata = anidataloader(file_path)
+        anidata_size = anidata.group_size()
+        enable_pkbar = anidata_size > 5 and PKBAR_INSTALLED
+        if enable_pkbar:
+            pbar = pkbar.Pbar('=> loading h5 dataset into cpu memory, total molecules: {}'.format(anidata_size), anidata_size)
+
+        for i, molecule in enumerate(anidata):
+            num_conformations = len(molecule['coordinates'])
+            # species and coordinates
+            self.data_coordinates += list(molecule['coordinates'].reshape(num_conformations, -1).astype(np.float32))
+            species = np.array([species_dict[x] for x in molecule['species']])
+            self.data_species += list(np.tile(species, (num_conformations, 1)))
+            # energies
+            self.data_energies += list(molecule['energies'].reshape((-1, 1)).astype(np.float32))
+            if subtract_self_energies:
+                self_energies = np.array(sum([self_energies_dict[x] for x in molecule['species']]))
+                self.data_self_energies += list(np.tile(self_energies, (num_conformations, 1)).astype(np.float32))
+            # other properties
+            if enable_pkbar:
+                pbar.update(i)
+
+        if subtract_self_energies:
+            self.data_energies = np.array(self.data_energies) - np.array(self.data_self_energies)
+            import gc
```

Comment:
`gc` is part of the standard library, so import it at the beginning of this file.